### ISSUE TITLE
select on /dev/sdm7 (GPS data)

### ISSUE THREAD

**Issue by nikitos1550** [Biktorgj/quectel_eg25_kernel#3]
Hello!

I am trying to handle gps data on mdm9607.

`echo -ne "AT+QGPS=1\r"` | microcom -t 100 /dev/smd9 for this kernel (or smd8 if I use original 3.18 downstream kernel).

After I can obtain data from /dev/smd7

```
# cat /dev/smd7
$GPVTG,,T,,M,,N,,K,N*2C
$GPGSA,A,1,,,,,,,,,,,,,,,*1E
$GPGGA,,,,,,0,,,,,,,,*66
$GPRMC,,V,,,,,,,,,,N*53
$GPVTG,,T,,M,,N,,K,N*2C
$GPGSA,A,1,,,,,,,,,,,,,,,*1E
$GPGGA,,,,,,0,,,,,,,,*66
$GPRMC,,V,,,,,,,,,,N*53
^C
```

But gpsd (as it is proper way to handle NMEA data) can't read from this tty device.
Seems it is because pselect/select for some reason doesn't work with 

```
[CUT]
openat(AT_FDCWD, "/etc/localtime", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
getuid32()                              = 0
stat64("/dev/smd7", {st_mode=S_IFCHR|0600, st_rdev=makedev(248, 7), ...}) = 0
chmod("/dev/smd7", 020660)              = 0
setgroups32(0, [])                      = 0
socket(PF_LOCAL, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 5
connect(5, {sa_family=AF_LOCAL, sun_path="/var/run/nscd/socket"}, 110) = -1 ENOENT (No such file or directory)
close(5)                                = 0
socket(PF_LOCAL, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 5
connect(5, {sa_family=AF_LOCAL, sun_path="/var/run/nscd/socket"}, 110) = -1 ENOENT (No such file or directory)
close(5)                                = 0
openat(AT_FDCWD, "/etc/group", O_RDONLY|O_CLOEXEC) = 5
fstat64(5, {st_mode=S_IFREG|0644, st_size=361, ...}) = 0
_llseek(5, 0, [0], SEEK_SET)            = 0
read(5, "root:x:0:\ndaemon:x:1:\nbin:x:2:\ns"..., 4096) = 361
close(5)                                = 0
setgid32(18)                            = 0
socket(PF_LOCAL, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 5
connect(5, {sa_family=AF_LOCAL, sun_path="/var/run/nscd/socket"}, 110) = -1 ENOENT (No such file or directory)
close(5)                                = 0
socket(PF_LOCAL, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0) = 5
connect(5, {sa_family=AF_LOCAL, sun_path="/var/run/nscd/socket"}, 110) = -1 ENOENT (No such file or directory)
close(5)                                = 0
openat(AT_FDCWD, "/etc/passwd", O_RDONLY|O_CLOEXEC) = 5
fstat64(5, {st_mode=S_IFREG|0644, st_size=395, ...}) = 0
_llseek(5, 0, [0], SEEK_SET)            = 0
read(5, "root:x:0:0:root:/root:/bin/sh\nda"..., 4096) = 395
close(5)                                = 0
setuid32(65534)                         = 0
rt_sigaction(SIGHUP, {0x15d7c, ~[RTMIN RT_1], SA_RESTORER, 0xb6c8a2e0}, NULL, 8) = 0
rt_sigaction(SIGINT, {0x15d7c, ~[RTMIN RT_1], SA_RESTORER, 0xb6c8a2e0}, NULL, 8) = 0
rt_sigaction(SIGTERM, {0x15d7c, ~[RTMIN RT_1], SA_RESTORER, 0xb6c8a2e0}, NULL, 8) = 0
rt_sigaction(SIGQUIT, {0x15d7c, ~[RTMIN RT_1], SA_RESTORER, 0xb6c8a2e0}, NULL, 8) = 0
rt_sigaction(SIGPIPE, {SIG_IGN, [PIPE], SA_RESTORER|SA_RESTART, 0xb6c8a2e0}, {SIG_DFL, [], 0}, 8) = 0
clock_gettime(CLOCK_REALTIME_COARSE, {5624, 999999915}) = 0
getpid()                                = 298
getpid()                                = 298
getsid(298)                             = 257
write(2, "gpsd:ERROR: system time looks bo"..., 64gpsd:ERROR: system time looks bogus, dates may not be reliable.
) = 64
pselect6(5, [3 4], NULL, NULL, {5, 0}, {NULL, 8}) = 0 (Timeout)
pselect6(5, [3 4], NULL, NULL, {5, 0}, {NULL, 8}) = 0 (Timeout)
pselect6(5, [3 4], NULL, NULL, {5, 0}, {NULL, 8}^CProcess 298 detached
 <detached ...>
```

I checked `/drivers/tty/serial/msm_smd_tty.c`, but I am not very familiar with deep kernel's internals.

Maybe you faced something similar before or just know what can I dig to debug this issue?

P.S.
```
# stty -aF /dev/smd7
speed 38400 baud;stty: /dev/smd7: No such file or directory
 line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>;
eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R;
werase = ^W; lnext = ^V; flush = ^O; min = 1; time = 0;
-parenb -parodd -cmspar cs8 -hupcl -cstopb cread -clocal -crtscts
-ignbrk -brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr -icrnl -ixon
-ixoff -iuclc -ixany -imaxbel -iutf8
-opost -olcuc -ocrnl -onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0
ff0
-isig -icanon -iexten -echo -echoe -echok -echonl -noflsh -xcase -tostop
-echoprt -echoctl -echoke -flusho -extproc
```
stty for some reason reports `/dev/smd7: No such file or directory`

**Comment by nikitos1550** [comment]
I found that gpsd uses pselect on the port, but https://github.com/Biktorgj/quectel_eg25_kernel/blob/linux-3.18.140/drivers/tty/serial/msm_smd_tty.c doesn't support poll. Seems it somehow can be added via tty_ldisc_ops. 
But at the moment, I have no idea how to do it.

**Comment by nikitos1550** [comment]
Sorry, I did wrong diagnostic for the issue. After second look on the strace log, I realized that gpsd doesn't open /dev/smd7 before there is some client (gpsmon for example), then I catch permission denied error (proper /dev/smd7 permissions should be setuped). Also I had wrong idea how tty serial Linux subsystem is orginized, poll is already implemented in line discipline and tty core use tty driver via line discipline.
